### PR TITLE
AdminSessionsManager and AdminSessionInfo - strtotime issue fix for admin login PR

### LINKED ISSUE #34415: M 2.4.3-p1 - TypeError in AdminSessionsManager.php:338 (strtotime() expects parameter 1 to be string, int given)
### Preconditions (*)

1. Magento 2.4.3-p1

### Steps to reproduce (*)

1. When I try to login with SSO...

### Expected result (*)

1. Login works

### Actual result (*)

```
TypeError: strtotime() expects parameter 1 to be string, int given
#28 /var/www/share/meevo.io/releases/66/vendor/magento/module-security/Model/AdminSessionsManager.php(338): strtotime
```

magento2/app/code/Magento/Security/Model/AdminSessionsManager.php 

```
private function lastProlongIsOldEnough()
    {
        $lastProlongTimestamp = strtotime($this->getCurrentSession()->getUpdatedAt());
        $nowTimestamp = $this->authSession->getUpdatedAt();

        $diff = $nowTimestamp - $lastProlongTimestamp;

        return (float) $diff > $this->getIntervalBetweenConsecutiveProlongs();
    }

```

In this file the line https://github.com/magento/magento2/compare/2.4.3...2.4.3-p1#diff-24fbc4a93c8c25f9ee036f3bf353c3dfca22ee7ff35d0797b6d1c1d7dc012386R6

declare(strict_types=1);

$this->getCurrentSession()->getUpdatedAt() returns int as timestamp... for example 1635087142

Here is the complete trace

```
TypeError: strtotime() expects parameter 1 to be string, int given
#28 /vendor/magento/module-security/Model/AdminSessionsManager.php(338): strtotime
#27 /vendor/magento/module-security/Model/AdminSessionsManager.php(338): Magento\Security\Model\AdminSessionsManager::lastProlongIsOldEnough
#26 /vendor/magento/module-security/Model/AdminSessionsManager.php(137): Magento\Security\Model\AdminSessionsManager::processProlong
#25 /vendor/magento/module-security/Model/Plugin/AuthSession.php(85): Magento\Security\Model\Plugin\AuthSession::aroundProlong
#24 /vendor/magento/framework/Interception/Interceptor.php(135): Magento\Backend\Model\Auth\Session\Interceptor::Magento\Framework\Interception\{closure}
#23 /vendor/magento/framework/Interception/Interceptor.php(153): Magento\Backend\Model\Auth\Session\Interceptor::___callPlugins
#22 /Magento/Backend/Model/Auth/Session/Interceptor.php(23): Magento\Backend\Model\Auth\Session\Interceptor::prolong
#21 /vendor/magento/module-backend/App/Action/Plugin/Authentication.php(127): Magento\Backend\App\Action\Plugin\Authentication::aroundDispatch
#20 /vendor/magento/framework/Interception/Interceptor.php(135): Magento\Backend\Controller\Adminhtml\Auth\Login\Interceptor::Magento\Framework\Interception\{closure}
#19 /vendor/magento/framework/Interception/Interceptor.php(153): Magento\Backend\Controller\Adminhtml\Auth\Login\Interceptor::___callPlugins
#18 /Magento/Backend/Controller/Adminhtml/Auth/Login/Interceptor.php(32): Magento\Backend\Controller\Adminhtml\Auth\Login\Interceptor::dispatch
#17 /vendor/magento/framework/App/FrontController.php(245): Magento\Framework\App\FrontController::getActionResponse
#16 /vendor/magento/framework/App/FrontController.php(212): Magento\Framework\App\FrontController::processRequest
#15 /vendor/magento/framework/App/FrontController.php(147): Magento\Framework\App\FrontController::dispatch
#14 /vendor/magento/framework/Interception/Interceptor.php(58): Magento\Framework\App\FrontController\Interceptor::___callParent
#13 /vendor/magento/framework/Interception/Interceptor.php(138): Magento\Framework\App\FrontController\Interceptor::Magento\Framework\Interception\{closure}
#12 /src/Infortis/ext-magento2-cgen/src/Plugin/Magento/Framework/App/FrontController.php(32): Infortis\Cgen\Plugin\Magento\Framework\App\FrontController::aroundDispatch
#11 /vendor/magento/framework/Interception/Interceptor.php(135): Magento\Framework\App\FrontController\Interceptor::Magento\Framework\Interception\{closure}
#10 /vendor/magento/framework/Interception/Interceptor.php(153): Magento\Framework\App\FrontController\Interceptor::___callPlugins
#9 /Magento/Framework/App/FrontController/Interceptor.php(23): Magento\Framework\App\FrontController\Interceptor::dispatch
#8 /vendor/magento/framework/App/Http.php(116): Magento\Framework\App\Http::launch
#7 /vendor/magento/framework/Interception/Interceptor.php(58): Magento\Framework\App\Http\Interceptor::___callParent
#6 /vendor/magento/framework/Interception/Interceptor.php(138): Magento\Framework\App\Http\Interceptor::Magento\Framework\Interception\{closure}
#5 /vendor/justbetter/magento2-sentry/Plugin/GlobalExceptionCatcher.php(58): JustBetter\Sentry\Plugin\GlobalExceptionCatcher::aroundLaunch
#4 /vendor/magento/framework/Interception/Interceptor.php(135): Magento\Framework\App\Http\Interceptor::Magento\Framework\Interception\{closure}
#3 /vendor/magento/framework/Interception/Interceptor.php(153): Magento\Framework\App\Http\Interceptor::___callPlugins
#2 /Magento/Framework/App/Http/Interceptor.php(23): Magento\Framework\App\Http\Interceptor::launch
#1 /vendor/magento/framework/App/Bootstrap.php(264): Magento\Framework\App\Bootstrap::run
#0 /index.php(29): null

```
---
Please provide [Severity](https://devdocs.magento.com/guides/v2.4/contributor-guide/contributing.html#backlog) assessment for the Issue as Reporter. This information will help during Confirmation and Issue triage processes.

- [ ] Severity: **S0** _- Affects critical data or functionality and leaves users without workaround._
- [x] Severity: **S1** _- Affects critical data or functionality and forces users to employ a workaround._
- [ ] Severity: **S2** _- Affects non-critical data or functionality and forces users to employ a workaround._
- [ ] Severity: **S3** _- Affects non-critical data or functionality and does not force users to employ a workaround._
- [ ] Severity: **S4** _- Affects aesthetics, professional look and feel, “quality” or “usability”._

### LINKED ISSUE #34531: Backend throws PHP Fatal TypeError when session storage is empty
**Description**
When the admin session storage is empty, a PHP Fatal TypeError occurs:
```
PHP Fatal error:  Uncaught TypeError: strtotime() expects parameter 1 to be string, null given in .../vendor/magento/module-security/Model/AdminSessionInfo.php:136
Stack trace:
#0 .../vendor/magento/module-security/Model/AdminSessionInfo.php(136): strtotime(NULL)
#1 .../vendor/magento/module-security/Model/AdminSessionInfo.php(119): Magento\Security\Model\AdminSessionInfo->isSessionExpired()
#2 .../vendor/magento/module-security/Model/AdminSessionInfo.php(108): Magento\Security\Model\AdminSessionInfo->checkActivity()
#3 .../vendor/magento/module-security/Model/Plugin/AuthSession.php(63): Magento\Security\Model\AdminSessionInfo->isLoggedInStatus()
#4 .../vendor/magento in .../vendor/magento/module-security/Model/AdminSessionInfo.php on line 136
```
Please note that line numbers might be different in the current version, the problem however still exists.

**Since**
_Introduced with MC-34197:_
https://github.com/magento/magento2/commit/67fae82c6278827f2c9b0f63083486a9d7740366
_Specific:_
https://github.com/magento/magento2/blob/6b196b5df43f63c7364f7e193709694abca576d0/app/code/Magento/Backend/Model/Auth/Session.php#L186

**Suggested Fix**
Return `true` if the session is blank:
https://github.com/magento/magento2/blob/2.4-develop/app/code/Magento/Security/Model/AdminSessionInfo.php#L130
```
    public function isSessionExpired()
    {
        $lifetime = $this->securityConfig->getAdminSessionLifetime();
        $currentTime = $this->dateTime->gmtTimestamp();
        $lastUpdatedTime = $this->getUpdatedAt();
        if (empty($lastUpdatedTime)) {
            return true;
        }
        if (!is_numeric($lastUpdatedTime)) {
            $lastUpdatedTime = strtotime($lastUpdatedTime);
        }

        return $lastUpdatedTime <= ($currentTime - $lifetime) ? true : false;
    }
```

### PR DESCRIPTION
**Pull request by kanhaiya5590**
### Description (*)

 - TypeError: strtotime() expects parameter 1 to be string, int given ...security/Model/AdminSessionsManager.php:338 
 - TypeError: strtotime() expects parameter 1 to be string, null given in ...security/Model/AdminSessionInfo.php:136

### Fixed Issues (if relevant)

1. Fixes magento/magento2#34415
2. Fixes https://github.com/magento/magento2/issues/34531

### Manual testing scenarios (*)

1. ...
2. ...

### Contribution checklist (*)
 - [ ] Pull request has a meaningful description of its purpose
 - [ ] All commits are accompanied by meaningful commit messages
 - [ ] All new or changed code is covered with unit/integration tests (if applicable)
 - [ ] README.md files for modified modules are updated and included in the pull request if any [README.md predefined sections](https://github.com/magento/devdocs/wiki/Magento-module-README.md) require an update
 - [ ] All automated tests passed successfully (all builds are green)

### CODE REVIEW

**Comment by kanhaiya5590** [comment]
@magento give me 2.4-develop instance

File: app/code/Magento/Security/Model/AdminSessionInfo.php [hunk 1]
```diff
@@ -133,7 +133,11 @@ public function isSessionExpired()
         $currentTime = $this->dateTime->gmtTimestamp();
         $lastUpdatedTime = $this->getUpdatedAt();
         if (!is_numeric($lastUpdatedTime)) {
-            $lastUpdatedTime = strtotime($lastUpdatedTime);
+            if($lastUpdatedTime === null) {
+                $lastUpdatedTime = 0;
+            }else {
```
Comment (Den4ik) [inline app/code/Magento/Security/Model/AdminSessionInfo.php]:
```suggestion
            } else {
```

**Comment by Den4ik** [comment]
@magento run all tests

**Comment by kanhaiya5590** [comment]
@magento run all tests

**Comment by Den4ik** [comment]
@magento run all tests

**Comment by hostep** [comment]
Watch out that there are 2 different PR's trying to fix the same issue:
- this one
- and: https://github.com/magento/magento2/pull/34520

Only one of both should be chosen, right?

**Comment by johnulecan** [comment]
Hey @hostep . I guess only one PR will be accepted. However, I think there is no need to initialize a timestamp with 0 and also there is no need to add additional conditions. The problem lies in AdminSessionInfo.php isSessionExpired() - wreturn bool - method. If the session is expired, getUpdatedAt returns null and strtotime expects string. This fixes the whole problem: https://patch-diff.githubusercontent.com/raw/magento/magento2/pull/34520.diff

**Comment by Den4ik** [comment]
Yep, only one PR could be accepted

File: app/code/Magento/Security/Model/AdminSessionInfo.php [hunk 2]
```diff
@@ -133,7 +134,11 @@ public function isSessionExpired()
         $currentTime = $this->dateTime->gmtTimestamp();
         $lastUpdatedTime = $this->getUpdatedAt();
         if (!is_numeric($lastUpdatedTime)) {
-            $lastUpdatedTime = strtotime($lastUpdatedTime);
+            if ($lastUpdatedTime === null) {
+                $lastUpdatedTime = 0;
+            } else {
+                $lastUpdatedTime = strtotime($lastUpdatedTime);
+            }
```
Comment (ihor-sviziev) [inline app/code/Magento/Security/Model/AdminSessionInfo.php]:
```suggestion
            $lastUpdatedTime = $lastUpdatedTime === null ? 0 : strtotime($lastUpdatedTime);
```

File: app/code/Magento/Security/Model/AdminSessionInfo.php [hunk 3]
```diff
@@ -133,7 +134,11 @@ public function isSessionExpired()
...
         }
 
         return $lastUpdatedTime <= ($currentTime - $lifetime) ? true : false;
```
Comment (ihor-sviziev) [inline app/code/Magento/Security/Model/AdminSessionInfo.php]:
```suggestion
        return $lastUpdatedTime <= ($currentTime - $lifetime);
```

**Review by ihor-sviziev — CHANGES_REQUESTED** [review]
Hi @kanhaiya5590,
Could you apply suggested changes and cover your changes with unit tests?

File: app/code/Magento/Security/Model/AdminSessionInfo.php (hunk 3 above)
Comment (kanhaiya5590) [reply app/code/Magento/Security/Model/AdminSessionInfo.php]: this remove make no sense but as return will be either true or false, to optimize point of view its good to remove

**Comment by kanhaiya5590** [comment]
Hi @hostep @johnulecan @ihor-sviziev @Den4ik ,

Can you help to add Test Cases for changes.

or @magento-engcom-team  help for unit test

**Comment by kanhaiya5590** [comment]
@magento run all tests

**Comment by kanhaiya5590** [comment]
@magento run Unit Tests

**Comment by kanhaiya5590** [comment]
> @kanhaiya5590 i've created 2 unit tests, maybe you can use them for your PR: [Tjitse-E@32a742e](https://github.com/Tjitse-E/magento2/commit/32a742eb5b00249f757df079ea373c975ed8d1de)

@Tjitse-E  Thanks for Unit tests - added in PR

**Comment by ihor-sviziev** [comment]
@magento run all tests

**Review by ihor-sviziev — CHANGES_REQUESTED** [review]
Hi @kanhaiya5590,
Now the changes look good to me.
Could you fix failing static tests and I'll approve your PR?

**Comment by kanhaiya5590** [comment]
@magento run Static Tests

**Comment by ihor-sviziev** [comment]
@kanhaiya5590 Great job!
@magento run all tests

**Review by ihor-sviziev — APPROVED** [review]

**Comment by magento-engcom-team** [comment]
Hi @ihor-sviziev, thank you for the review.
[ENGCOM-9336](https://jira.corp.magento.com/browse/ENGCOM-9336) has been created to process this Pull Request

**Comment by lytesaber** [comment]
I see this is being merged into 2.4-develop branch however this problem also the affects 2.3.7-p2 release are we going to get a patch for the 2.3.x release line as well? Currently we're having to ask clients to clear their cookies to be able to access the admin panel of their stores due to this error.

**Comment by ihor-sviziev** [comment]
@magento run all tests

**Comment by ihor-sviziev** [comment]
@sidolov, how can we convert this PR to a patch to 2.3.7-p2?

**Comment by kanhaiya5590** [comment]
HI @ihor-sviziev 

Every pull-request on GH can be downloaded as a beautiful mail-patch, just by appending ".patch" to the PR URL

https://patch-diff.githubusercontent.com/raw/magento/magento2/pull/34514.patch 

This will generate a mail-formatted patch file, that is a little different from an usual patch file: It has e-mail metadata.

After downloaded, apply the result patch using:
`$ git am 34514.patch`

Hope this is what patch you looking for.

or PR diff -> https://patch-diff.githubusercontent.com/raw/magento/magento2/pull/34514.diff

**Comment by Den4ik** [comment]
> @sidolov, how can we convert this PR to a patch to 2.3.7-p2?

Similar as cookie patch. I made PR manually. @kanhaiya5590 Could you prepare PR to 2.3-release branch?

**Review by Den4ik — APPROVED** [review]

**Comment by magento-engcom-team** [comment]
Hi @Den4ik, thank you for the review.
[ENGCOM-9336](https://jira.corp.magento.com/browse/ENGCOM-9336) has been created to process this Pull Request

**Comment by kanhaiya5590** [comment]
Hi @Den4ik 
Created  a PR for Magento v2.3 https://github.com/magento/magento2/pull/34606

**Comment by ihor-sviziev** [comment]
@kanhaiya5590, yeah, I know, the question was about creating the quality patch for magento 2.3.7-p2

**Comment by ihor-sviziev** [comment]
Hi @kanhaiya5590,
It looks like this issue isn't reproducing on clean Magento installation. Could you provide some steps to reproduce it?

**Comment by DannyDaemonic** [comment]
I just ran into issue #34415. Applying this PR (#34514) resolved the problem.

@ihor-sviziev I apologize if this isn't productive, as I can't say the exact steps, but it seems it was at least partially triggered by logging in from a different system and ip address without first logging out of my previous session. When I went back to the system I was previously logged in to, I ran into the issue.

**Comment by engcom-Alfa** [comment]
:heavy_check_mark: QA Passed

**Preconditions:**

1. Have Magento latest installed


**Manual testing scenario:**

1. Update the code as shown in below screenshot.
Screenshot 2021-12-27 at 2 29 44 PM

2. Try to login as an Admin from UI.


**Before:** :heavy_multiplication_x: Getting an typeError:

Screenshot 2021-12-27 at 2 29 55 PM


**After:** :heavy_check_mark: Able to see the values in the response 


Since it is a specific case were code changes are involved, no additional regression is required on this!

**Comment by kanhaiya5590** [comment]
Hope this PR is ready to merge.

FYi - @ihor-sviziev @engcom-Alfa @Den4ik

**Pull request merged**